observability/metrics/src/test/lib/scala/com/daml/metrics/api/testing/InMemoryMetricsFactory.scala

```diff
@@ -21,48 +21,106 @@ import com.daml.metrics.api.testing.InMemoryMetricsFactory.{
   InMemoryHistogram,
   InMemoryMeter,
   InMemoryTimer,
+  MetricsState,
 }
 import com.daml.metrics.api.{MetricHandle, MetricName, MetricsContext}
 import com.daml.scalautil.Statement.discard
 
 import scala.collection.concurrent.{TrieMap, Map => ConcurrentMap}
+import scala.collection.mutable
 
 class InMemoryMetricsFactory extends LabeledMetricsFactory {
 
-  val asyncGauges: ConcurrentMap[(MetricName, MetricsContext), () => Any] =
-    TrieMap[(MetricName, MetricsContext), () => Any]()
+  val metrics: MetricsState =
+    MetricsState(
+      timers = mutable.Map.empty,
+      gauges = mutable.Map.empty,
+      meters = mutable.Map.empty,
+      counters = mutable.Map.empty,
+      histograms = mutable.Map.empty,
+      asyncGauges = TrieMap[(MetricName, MetricsContext), () => Any](),
+    )
 
   override def timer(name: MetricName, description: String)(implicit
       context: MetricsContext
-  ): MetricHandle.Timer = InMemoryTimer(context)
+  ): MetricHandle.Timer = metrics.addTimer(name, context, InMemoryTimer(context))
 
   override def gauge[T](name: MetricName, initial: T, description: String)(implicit
       context: MetricsContext
-  ): MetricHandle.Gauge[T] = InMemoryGauge(context, initial)
+  ): MetricHandle.Gauge[T] = metrics.addGauge(name, context, InMemoryGauge(context, initial))
 
   override def gaugeWithSupplier[T](
       name: MetricName,
       gaugeSupplier: () => T,
       description: String,
-  )(implicit context: MetricsContext): Unit = asyncGauges.addOne(name -> context -> gaugeSupplier)
+  )(implicit context: MetricsContext): Unit =
+    metrics.asyncGauges.addOne(name -> context -> gaugeSupplier)
```

Comment:
[No action required] I understand that it is not really needed, but it is a bit strange that gaugeWithSupplier doesn't follow the pattern of the other metrics: a `GaugeWithSupplier` class, and a `metrics.addGaugeWithSupplier` method.